Forward fix internal pyre failure from D58983461

huydhn opened the pull request:
Summary: Somehow, using underscore alias of some builtin types breaks pyre

Test Plan:
All failed tests from D58983461 are passing:

```
buck2 test 'fbcode//mode/opt' fbcode//caffe2/torch/fb/training_toolkit/utils/tests:gpu_memory_utils_test-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/lib:device_util-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/lib:thompson_samplers_gpu-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/modules/retrieval/diversity/tests:combined_sampling_diversifier_test-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/modules/retrieval/diversity/tests:submodular_opt_test-type-checking
```

Differential Revision: D59029768

huydhn commented:
@XuehaiPan This seems to pass all Meta-internal pyre check from your PR https://github.com/pytorch/pytorch/pull/129001.

@malfet Please help take a look and let me know if this is a reasonable forward fix (as it looks a bit weird from what I see but I couldn't really explain it)

XuehaiPan commented on torch/types.py:
```diff
@@ -36,16 +38,16 @@
 _layout = torch.layout
-_size = Union[torch.Size, List[_int], Tuple[_int, ...]]
-_dispatchkey = Union[_str, torch._C.DispatchKey]
+_size = Union[torch.Size, List[builtins.int], Tuple[builtins.int, ...]]
```
I wonder if it works without `builtins.` namespace.

```suggestion
_size = Union[torch.Size, List[int], Tuple[int, ...]]
```

huydhn replied:
Let me try it out, I think that pyre doesn't run the code but only does static code analysis on top and probably confuse `_int` with a variable.

malfet reviewed (changes requested):
But this is wrong, because `int` here is actually `torch.int`, which is an alias to `torch.int32`, while `_int` is an alias to `int` type

huydhn commented:
> But this is wrong, because `int` here is actually `torch.int`, which is an alias to `torch.int32`, while `_int` is an alias to `int` type

I see.  Do you think using `builtins.int` explicitly works?  At least, it satisfies pyre from my testing

huydhn commented:
@malfet I bite the bullet and will land this with `builtins.int` to unblock today train, if that change still has issue, plz feel free to revert both this and https://github.com/pytorch/pytorch/pull/129001

malfet commented:
Using builtins as a short term workaround sounds good to me, but let's add some sort of a followup issue to investigate what has happened, as we've been able to use aliases since the dawn of times.

huydhn committed:
Forward fix internal pyre failure from D58983461 (#129525)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/129525

Somehow, using underscore alias of some builtin types breaks pyre

Test Plan:
All failed tests from D58983461 are passing, also run `arc lint`

```
buck2 test 'fbcode//mode/opt' fbcode//caffe2/torch/fb/training_toolkit/utils/tests:gpu_memory_utils_test-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/lib:device_util-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/lib:thompson_samplers_gpu-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/modules/retrieval/diversity/tests:combined_sampling_diversifier_test-type-checking
buck2 test 'fbcode//mode/dev-nosan' fbcode//dper_lib/silvertorch/modules/retrieval/diversity/tests:submodular_opt_test-type-checking
```

Reviewed By: malfet, clee2000

Differential Revision: D59029768

XuehaiPan commented:
> But this is wrong, because `int` here is actually `torch.int`, which is an alias to `torch.int32`, while `_int` is an alias to `int` type

In `torch/__init__.py`, we must use `builtins.int`, `buultins.float`, etc. But we are annotating `torch/types.py`, the builtins classes are not shadowed.
